### PR TITLE
[MOB-2734] Implement Snapshot Testing

### PR DESCRIPTION
**Pull request by d4r1091**
[MOB-2734]

## Context

We want to enable snapshot testing on the Ecosia iOS project.
Our north ⭐ is: We want to [KISS](https://en.wikipedia.org/wiki/KISS_principle) to make snapshot across different devices, languages and so on. 

## Approach

Snapshot Testing Infrastructure Implementation

We’ve introduced a snapshot testing setup using the SnapshotTesting library, aimed at maintaining visual consistency in our UI across various updates. This is crucial as we scale up our app’s features and interfaces.

### Implementation Details

Our implementation revolves around the SnapshotTestHelper class designed to streamline the setup of snapshot tests. It supports a variety of themes, devices, and locales. Notably, we added functionality to dynamically switch locales and apply themes during tests, which enhances the accuracy and coverage of our UI tests.

`EcosiaTests/SnapshotTests/SnapshotTestHelper.swift` is the core of the implementation.

### Challenges

One significant challenge was the implementation of dynamic theming (due to the current issue the vanilla codebase suffer from when swapping dark/light mode) and locale switching. 
Our app didn’t support in-app language changes, so we had to engineer a workaround that involved overriding system locales and refreshing views. This required careful manipulation of app settings and bundle resources, which was a complex but rewarding problem to solve.

### Repository Size Considerations

We’re aware that adding snapshot images for multiple devices and locales will increase our repository size. To manage this efficiently, we’re planning to move the snapshot images to a separate submodule. This will keep our main repository clean while allowing easy access to test assets when needed by our CI systems.

### Future Steps

- CI Integration: The immediate next step is to integrate this testing setup with our CI pipeline to ensure every commit is automatically checked for UI consistency.
- Submodule Setup: We’ll be setting up a follow-up ticket to explore creating a submodule for our snapshot images. This will help us manage repository size and streamline updates to our visual test assets.

This setup not only enhances our testing capabilities but also prepares us to handle more complex UI scenarios as we continue to evolve our application’s interface.

## Other

- With this SnapshotTesting work we already spotted some theming issue that we promptly tackled.
- We also implemented a shortcut that would let us move to the correct onboarding screen without having to go back and forth.
- The`NTPTests` are there but deactivated (therefore the need of a TestPlan) due to an issue with the async favicons load that will stop the execution prematurely. We can continue investigate it but for now it's not so relevant as we are testing the single components of that screen anyway.
- We are also refreshing our bundle deps
- I will close this PR https://github.com/ecosia/ios-browser/pull/722 in favour of the reviewed implementation

### Checklist

- [x] I performed some relevant testing on a real device and/or simulator
- [x] I wrote Unit Tests that confirm the expected behaviour
- [x] I added the `// Ecosia:` helper comments where needed
- [x] I included documentation updates to the coding standards or Confluence doc, when needed

[MOB-2734]: https://ecosia.atlassian.net/browse/MOB-2734?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: EcosiaTests/SnapshotTests/Onboarding/OnboardingTests.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    }
+    
+    func testWelcomeStepsScreens() {
+        // Number of steps in the WelcomeTour
+        let numberOfSteps = 4
+        // Iterate through steps and take snapshots, skipping the first one
+        for step in 1...numberOfSteps {
+            let startingStep = WelcomeTour.Step.all[step-1]
+            // Precision at .95 to accommodate a snapshot looking slightly different due to the different data output
+            // from the statistics json
```
Comment (lucaschifino) [inline EcosiaTests/SnapshotTests/Onboarding/OnboardingTests.swift:24]: Did you look into mocking the json? Not major but might be interesting to test how different statistics look as well

File: SNAPSHOT_TESTING_WIKI.md
Comment (lucaschifino) [inline SNAPSHOT_TESTING_WIKI.md:1]: 🤩

**Review by lucaschifino — APPROVED** [review]
It's looking great, awesome job with the mocks and the docs 🔥 

Also like the idea of using submodules 💡 Would also be interesting if we made the images accessible to product so out snapshotted UI can serve as a source of truth of the screens.

File: EcosiaTests/SnapshotTests/Onboarding/OnboardingTests.swift (hunk 1 above)
Comment (d4r1091) [reply EcosiaTests/SnapshotTests/Onboarding/OnboardingTests.swift:24]: Yes we'd mock it but that would require further work. Would like to tackle it separately

**Pull request merged**